VM release improvements

josephglanville opened the pull request:
Builds on #1718 to address #1719 

Still somewhat WIP and requires more testing and probably work to make it conform to bash styleguide.
But the general idea is there.

Still to go is QEMU support and verification that the publishing side is not bug riddled.

josephglanville commented:
@lmars if you could take a look at this that would be great.

I am probably not going to get time to look at libvirt/qemu based stuff until the weekend so maybe do that as separate PR.

josephglanville commented:
While I am here there is another problem with this currently. It stores the stage1 artifacts in the same temporary directory as the final build products, meaning they are erased if the upload is successful.

How does adding another argument that takes a path to be used as a cache directory for the stage1 VM images sound?

lmars commented on script/release-vm-images:
```diff
@@ -44,19 +48,20 @@ usage: $0 [options] VERSION
+  -b BUCKET     The S3 bucket to upload vagrant images to [default: 'flynn']
+  -d DOMAIN     The CloudFront domain [default: 'dl.flynn.io']
   -r DIR        Resume the release using DIR
+  -c CACHE      Directory to cache base VM images [default: '/tmp']
```
It doesn't look like you are actually setting the default value of `cache`, I would expect to see the following somewhere

```bash
cache="${cache:-"/tmp"}"
```

josephglanville replied:
Good catch.

Initialised it to "/tmp" rather than creating the unset local.

lmars commented on script/release-vm-images:
```diff
@@ -86,46 +97,89 @@ main() {
   check_aws_keys
 
+  # Assume us-east-1 if not set.
+  AWS_DEFAULT_REGION=${AWS_DEFAULT_REGION:-us-east-1}
```
This should be a flag like the other optional variables.

josephglanville replied:
Done.

lmars commented on script/release-vm-images:
```diff
@@ -86,46 +97,89 @@ main() {
+  local stage1_imgs=(${dir}/base_*_*)
+  local stage1_targets=()
+  for builder in ${builders[@]}; do
+    if [[ ! "${stage1_imgs[@]}" =~ "base_${builder}_${UBUNTU_ISO_SHA}" ]]; then
```
I would drop the `stage1_imgs` variable and use `test -f`:


```bash
if [[ ! -f "${dir}/base_${builder}_${UBUNTU_ISO_SHA}" ]]; then
  stage1_targets+=("${builder}-iso")
fi
```

josephglanville replied:
Yeah, that is much nicer.

lmars commented on script/release-vm-images:
```diff
@@ -86,46 +97,89 @@ main() {
-    if [[ "$(sha256 "${ubuntu_box}")" != "${UBUNTU_BOX_SHA}" ]]; then
-      rm -f "${ubuntu_box}"
-      fail "base Ubuntu box has incorrect checksum"
+  if [ ${#stage1_targets[@]} -eq 0 ]; then
```
Use `[[`

josephglanville replied:
Done.

lmars commented on script/release-vm-images:
```diff
@@ -86,46 +97,89 @@ main() {
+  local boxes=(${dir}/flynn-base_*_*.box)
+  local stage2_targets=()
+  for builder in ${builders[@]}; do
+    if [[ ! "${boxes[@]}" =~ "flynn-base_${version}_${builder}.box" ]]; then
```
Again I would just use `test -f`

josephglanville replied:
Done.

josephglanville commented:
@lmars I have addressed comments and added another trivial change to configure the amount of memory/vcpus the VMs use. My build server isn't as chunky so I wanted to scale them down a little heh.

titanous commented:
Closing until this is revisited.
